### PR TITLE
feat(gatsby-source-graphql): Query batching

### LINKED ISSUE #13425: gatsby-source-graphql: n+1 queries to remote GraphQL APIs, and other concerns
In working on the initial phase of gatsbywpthemes.com, using WPGraphQL to pull data from WordPress into Gatsby, we initially used `gatsby-source-graphql` and stitched in the WPGraphQL Schema. 

Then the createPages query queried all the pages needed from WordPress asking for just the ID, passed that through then each page used a `pageQuery` to query the data needed to produce that page. 

One of the big benefits of using WPGraphQL is the reduction of n+1 queries when fetching data from WP, but by using this method, it was CREATING a new n+1 problem. 

If we had 1000 items in WP, we could execute 10 fetches for 100 items each, then the pageQuery would fetch once per page again, leading to a total of 1010 queries and slower builds. 

Our solution was to change to fetching the GraphQL API directly up front and passing data through as context. This allows us to make just the 10 up front requests, not 1,010 requests. 

However, readability becomes a concern because the `pageQuery` isn't coupled with the component anymore. 

I tried to solve that by having a fragment at the page level and importing the fragment into the query that runs at build time, but I was unable to acheive that. 

I have a Spectrum thread about this here: https://t.co/qeHM10MMHc

Related Twitter thread here: https://twitter.com/wpgraphql/status/1118570506750496772

**Also, related**

Another issue with `gatsby-source-graphql` (or at least my understanding of how to best use it) is that a stitched Query isn't meaninful to dev tools on the live API. 

You can't just copy a stitched query and execute it in GraphiQL against the live server. Stitched queries only have meaning in the Gatsby build GraphQL server, so if you're building a Gatsby site that also needs to pull data from a live WPGraphQL (or other live GraphQL) endpoint, you can't re-use queries across the build and dynamic fetching, you can't use the stitched queries in dev tooling on the live API, etc. 

Turns into a headache, in my experience. 

Ideally, I'd like to be able to re-use fragments from build on the client, but since Types are stitched, the build fragments won't work in live queries on the client. 

For example, the build query might look like:

```
query GET_PAGES {
  wpgraphql {
      pages {
         nodes {
            id
            title
            parent {
                __typename 
                ...on WPGraphQL_Page {
                  id
                  title
                }
            }
         }
      }
  }
}
```

You can't just use that on your `wordpress-site.com/graphql` endpoint because of the stitching. The live endpoint wouldn't understand it. Instead, it would expect the following query: 

```
query GET_PAGES {
  pages {
     nodes {
        id
        title
        parent {
           __typename
           ...on Page {
              id
              title
           }
        }  
     }
  }
}
```

Subtle differences, but `Page` vs `WordPress_Page`, and `{ pages{ ...} }` vs. `{ wpgraphql { pages { ... } } }` make it impossible to re-use fragments across client and build.

### PR DESCRIPTION
**Pull request by vladar**
## Description

This PR allows `gatsby-source-graphql` users to improve performance 2-10 times by introducing query batching strategy along with a new environment variable for controlling query concurrency.

By default, `gatsby-source-graphql` executes each query in a separate network request. Query batching strategy implemented in this PR merges several queries into a single query and sends it to the server.

Consider the following queries:

```js
{
  query: `query(id: Int!) {
    node(id: $id) {
      foo
    }
  }`,
  variables: { id: 1 },
}
```

```js
{
  query: `query(id: Int!) {
    node(id: $id) {
      bar
    }
  }`,
  variables: { id: 2 },
}
```

They will be merged into a single query:

```js
{
  query: `
    query(
      $gatsby0_id: Int!
      $gatsby1_id: Int!
    ) {
      gatsby0_node: node(id: $gatsby0_id) {
        foo
      }
      gatsby1_node: node(id: $gatsby1_id) {
        bar
      }
    }
  `,
  variables: {
    gatsby0_id: 1,
    gatsby1_id: 2,
  }
}
```

Then the response will be transformed back to the shape expected for original queries.

**Caveat**: Batching is only possible for queries starting approximately at the same time. In other words, it is bounded by the number of parallel GraphQL queries executed by Gatsby (by default it is **4**).

To work around this limitation this PR also introduces a new environment variable `GATSBY_EXPERIMENTAL_QUERY_CONCURRENCY`.

By tweaking this concurrency level and number of queries in batch it possible to significantly improve the performance in some cases.

Bench on [my test site](https://github.com/vladar/source-graphql-bench) (queries per second, bigger is better):

|                              |    no batching    | 5 queries per batch| 10 queries per batch |
|------------------------------|-------------------|--------------------|--------------------|
| Concurrency: 4 (baseline)    |      5 q/s        |    4.8 q/s         |       4.8 q/s      |
| Concurrency: 20              |      18.95 q/s    |    19.07 q/s       |       22.73 q/s    |
| Concurrency: 50              |      15 q/s       |    40.3 q/s        |       45 q/s       |
| Concurrency: 100             |  server died :)   |    44.3 q/s        |       48 q/s       |

### Documentation

The documentation is outlined in the README.md (happy for any advice)

## Related Issues
Fixes #13425

### CODE REVIEW

File: packages/gatsby-source-graphql/README.md [hunk 1]
```diff
@@ -176,5 +176,162 @@ module.exports = {
 }
 ```
 
+# Performance tuning
+
+By default, `gatsby-source-graphql` executes each query in a separate network request.
+But the plugin also supports query batching to improve query running performance
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
But the plugin also supports query batching to improve query performance.
```

File: packages/gatsby-source-graphql/README.md [hunk 2]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+
+**Caveat**: Batching is only possible for queries starting approximately at the same time. In other words
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
**Caveat**: Batching is only possible for queries starting at approximately the same time. In other words
```

File: packages/gatsby-source-graphql/README.md [hunk 3]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+it is bounded by the number of parallel GraphQL queries executed by Gatsby (by default it is **4**).
+
+Fortunately we can increase the number of queries executed in parallel by setting [environment variable](https://gatsby.dev/env-vars)
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
Fortunately, we can increase the number of queries executed in parallel by setting the [environment variable](https://gatsby.dev/env-vars)
```

File: packages/gatsby-source-graphql/README.md [hunk 4]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+`GATSBY_EXPERIMENTAL_QUERY_CONCURRENCY` to a higher value and setting `batch` option of the plugin
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
`GATSBY_EXPERIMENTAL_QUERY_CONCURRENCY` to a higher value and setting the `batch` option of the plugin
```

File: packages/gatsby-source-graphql/README.md [hunk 5]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+to `true`.
+
+Example:
+
+```
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
```shell
```

File: packages/gatsby-source-graphql/README.md [hunk 6]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+        url: "https://api.graphcms.com/simple/v1/swapi",
+        batch: true,
+      },
+    },
+  ],
+}
+```
+
+By default, the plugin batches up to 5 queries. You can override this by passing
+`dataLoaderOptions`:
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
`dataLoaderOptions` and set a `maxBatchSize`:
```

File: packages/gatsby-source-graphql/README.md [hunk 7]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+Each project is unique so try tuning those two variables and see what works best for you.
+We've seen up to 5-10x speed-up for some setups.
+
+### How batching works
+
+Under the hood `gatsby-source-graphql` uses [DataLoader](https://github.com/graphql/dataloader)
+for query batching. It merges all queries from a batch to a single query that gets sent to the
+server in a single network request.
+
+Consider following example:
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
Consider the following example where both of these queries are run:
```

File: packages/gatsby-source-graphql/README.md [hunk 8]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+
+```js
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
```graphql
```

File: packages/gatsby-source-graphql/README.md [hunk 9]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+  query: `query(id: Int!) {
+    node(id: $id) {
+      foo
+    }
+  }`,
+  variables: { id: 1 },
+}
+```
+
+```js
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
```graphql
```

File: packages/gatsby-source-graphql/README.md [hunk 10]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+  query: `query(id: Int!) {
+    node(id: $id) {
+      bar
+    }
+  }`,
+  variables: { id: 2 },
+}
+```
+
+They will be merged to a single query:
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
They will be merged into a single query:
```

File: packages/gatsby-source-graphql/README.md [hunk 11]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+
+```js
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
```graphql
```

File: packages/gatsby-source-graphql/README.md [hunk 12]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+    }
+  `,
+  variables: {
+    gatsby0_id: 1,
+    gatsby1_id: 2,
+  }
+}
+```
+
+Then `gatsby-source-graphql` splits the result of this single query to multiple results
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
Then `gatsby-source-graphql` splits the result of this single query into multiple results
```

File: packages/gatsby-source-graphql/README.md [hunk 13]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+and delivers it back to Gatsby as if it executed multiple queries:
+
+```js
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
```json
```

File: packages/gatsby-source-graphql/README.md [hunk 14]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+  data: {
+    gatsby0_node: { foo: `foo` },
+    gatsby1_node: { bar: `bar` },
+  },
+}
+```
+
+is transformed back to:
+
+```js
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
```json
```

File: packages/gatsby-source-graphql/README.md [hunk 15]
```diff
@@ -176,5 +176,162 @@ module.exports = {
...
+]
+```
+
+Note that if any query result contains errors the whole batch will fail.
+
+### Apollo-style batching
+
+If your server supports apollo-style query batching you can also try
+[HttpLinkDataLoader](https://github.com/prisma-labs/http-link-dataloader).
+Just pass it to the `gatsby-source-graphql` plugin via `createLink` option.
```
Comment (laurieontech) [inline packages/gatsby-source-graphql/README.md]:
```suggestion
Pass it to the `gatsby-source-graphql` plugin via the `createLink` option.
```

File: packages/gatsby-source-graphql/src/batching/dataloader-link.ts [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    if (!isValidGraphQLResult(result)) {
+      const error = extractFirstError(result)
+      throw new Error(`Failed to load query batch: ${error}`)
+    }
+    return resolveResult(result)
+  }
+
+  const dataloader = new DataLoader(load, {
+    cache: false,
+    maxBatchSize: 5,
```
Comment (freiksenet) [inline packages/gatsby-source-graphql/src/batching/dataloader-link.ts]: I'm thinking that maybe we should default to current max concurrency as maxBatchSize. Thoughts?

File: packages/gatsby-source-graphql/src/batching/dataloader-link.ts [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  headers?: object
+}
+
+export function createDataloaderLink(options: IOptions): ApolloLink {
+  const load = async (keys: ReadonlyArray<IQuery>): Promise<IQueryResult[]> => {
+    const query = merge(keys)
+    const result: object = await request(query, options)
+    if (!isValidGraphQLResult(result)) {
+      const error = extractFirstError(result)
+      throw new Error(`Failed to load query batch: ${error}`)
```
Comment (freiksenet) [inline packages/gatsby-source-graphql/src/batching/dataloader-link.ts]: Why only first error? I'd rather format them all (or are there too many?)

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+  parseKey: (prefixedKey: string): { index: number; originalKey: string } => {
+    const match = /^gatsby([\d]+)_(.*)$/.exec(prefixedKey)
+    if (!match || match.length !== 3 || isNaN(Number(match[1])) || !match[2]) {
+      throw new Error(`Unexpected data key: ${prefixedKey}`)
+    }
+    return { index: Number(match[1]), originalKey: match[2] }
+  },
+}
+
+/**
```
Comment (freiksenet) [inline packages/gatsby-source-graphql/src/batching/merge-queries.ts]: Nice.

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+
+    prefixedQuery.query.definitions.forEach(def => {
+      if (isQueryDefinition(def)) {
+        mergedSelections.push(...def.selectionSet.selections)
+        mergedVariableDefinitions.push(...(def.variableDefinitions ?? []))
+      }
+      if (isFragmentDefinition(def)) {
+        // Theoretically it is possible to have fragments with the same name but different content
+        // in different queries. But Gatsby validation doesn't allow this and we rely on this here
+        mergedFragmentMap.set(def.name.value, def)
```
Comment (freiksenet) [inline packages/gatsby-source-graphql/src/batching/merge-queries.ts]: This is probably okay, but *might* break if people use it in gatsby-node and have same fragments. I'm okay with adding a comment about that particular potential (unlikely) breakage.

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+  document = visit(
+    document,
+    visitInParallel([
+      // Note: the sequence is important due to how visitInParallel deals with node edits
+      Visitors.detectFragmentsWithVariables(fragmentsWithVariables),
+      Visitors.prefixVariables(prefix),
+    ])
+  )
+
+  if (fragmentsWithVariables.size > 0) {
```
Comment (freiksenet) [inline packages/gatsby-source-graphql/src/batching/merge-queries.ts]: Is prefixing fragments very expensive? Maybe we should prefix all fragments by default instead of doing two passes?

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+    switch (selection.kind) {
+      case Kind.INLINE_FRAGMENT:
+        return [aliasFieldsInInlineFragment(prefix, selection, document)]
+
+      case Kind.FRAGMENT_SPREAD: {
+        const inlineFragment = inlineFragmentSpread(selection, document)
+        return [
+          addSkipDirective(selection),
+          aliasFieldsInInlineFragment(prefix, inlineFragment, document),
+          // Keep original spread in selection with @skip(if: true)
```
Comment (freiksenet) [inline packages/gatsby-source-graphql/src/batching/merge-queries.ts]: Nifty!

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+  return {
+    ...fragment,
+    selectionSet: {
+      ...fragment.selectionSet,
+      selections: aliasFieldsInSelection(prefix, selections, document),
+    },
+  }
+}
+
+/**
```
Comment (freiksenet) [inline packages/gatsby-source-graphql/src/batching/merge-queries.ts]: One alternative would be to prefix by fragment also (like you do with query), but it might be harder to deconstruct later.

File: packages/gatsby-source-graphql/src/batching/dataloader-link.ts (hunk 16 above)
Comment (vladar) [reply packages/gatsby-source-graphql/src/batching/dataloader-link.ts]: So the thing is - concurrency gives more performance benefits than batching. Making `maxBatchSize` equal to concurrency limit means that we send a single query sequentially - this works roughly the same as sending queries concurrently.

Batching starts shining when you hit the concurrency limits of your server. Another possible issue with batching is that huge batched queries may overload servers or just fail the complexity scoring.

So to recap we can do `maxBatchSize = concurrencyLimit / optimialConcurrencyLevel`

Unfortunately `optimialConcurrencyLevel` will be different for different servers. We can probably stay on a safer side and set it to 4 or something like that.

File: packages/gatsby-source-graphql/src/batching/dataloader-link.ts (hunk 17 above)
Comment (vladar) [reply packages/gatsby-source-graphql/src/batching/dataloader-link.ts]: Good point. I will change this.

File: packages/gatsby-source-graphql/src/batching/dataloader-link.ts (hunk 16 above)
Comment (freiksenet) [reply packages/gatsby-source-graphql/src/batching/dataloader-link.ts]: Sounds good to me, let's do smth like that.

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts (hunk 20 above)
Comment (vladar) [reply packages/gatsby-source-graphql/src/batching/merge-queries.ts]: I've been thinking about this. But I suspect that fragments with variables will be **very** rare in the real world.

At the same time, some fragments can be huge (I've seen sites like that) which means that servers will spend more time parsing and validating them.

File: packages/gatsby-source-graphql/README.md (hunk 8 above)
Comment (vladar) [reply packages/gatsby-source-graphql/README.md]: This is actually a JS code example (object with query and variables).

File: packages/gatsby-source-graphql/README.md (hunk 9 above)
Comment (vladar) [reply packages/gatsby-source-graphql/README.md]: Also a JS example (graphql query is a part of JS object here)

File: packages/gatsby-source-graphql/README.md (hunk 11 above)
Comment (vladar) [reply packages/gatsby-source-graphql/README.md]: Also a JS example

File: packages/gatsby-source-graphql/README.md (hunk 13 above)
Comment (vladar) [reply packages/gatsby-source-graphql/README.md]: At the moment this example is a valid JS object but not a valid JSON. I guess I can change it to JSON but I thought it is more readable this way?

```js
{
  data: {
    gatsby0_node: { foo: `foo` },
    gatsby1_node: { bar: `bar` },
  },
}
```
vs
```json
{
  "data": {
    "gatsby0_node": { "foo": "foo" },
    "gatsby1_node": { "bar": "bar" }
  }
}
```
What do you think?

**Comment by laurieontech** [comment]
I'd actually prefer that we remove the object wrapper from those queries to make it a valid graphql query. It'll help clarify things since otherwise everything is a js tag. I'm mixed on the js vs. json. I'll let you pick which you prefer.

**Comment by vladar** [comment]
@laurieontech So GraphQL request is sent to the server as JSON:

```json
{
  "query": "query ($id: Int!) { foo(id: $id) }",
  "variables": {
    "id": "id",
  }
}
```

And batching transforms both the query and the variables. So we must somehow reflect it in the example. That's why we have this wrapper vs just graphql query.

P.S. Thanks a lot for your feedback on docs 🌮

**Comment by laurieontech** [comment]
Ah, that makes sense. Hadn't thought about the transformation piece. In that case I'd go for JS as well. So keep it as is :)

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts (hunk 22 above)
Comment (vladar) [reply packages/gatsby-source-graphql/src/batching/merge-queries.ts]: Yeah, that's a part of the tradeoff - prefixing all fragments or trying to de-duplicate and re-use fragments (without variables). We can reconsider if this proves to be a wrong decision, but for now, I'd prefer to keep it as is.

File: packages/gatsby-source-graphql/src/batching/merge-queries.ts (hunk 19 above)
Comment (vladar) [reply packages/gatsby-source-graphql/src/batching/merge-queries.ts]: Added this example to the comment

**Comment by vladar** [comment]
Published in `gatsby@2.20.5` and `gatsby-source-graphql@2.3.0`

**Comment by yaacovCR** [comment]
@vladar, any chance this batching link can be broken off into a separate package, even included within graphql-tools?

Would be helpful to community, I think.

**Comment by vladar** [comment]
The problem is that some parts of this code rely on Gatsby validations for simplicity. So it may fail in several edge cases outside of Gatsby. It is possible to make it work everywhere, it just requires time.

I think a good idea would be to wait for feedback from Gatsby users and then move this to a separate package (or `graphql-tools-fork`).

**Comment by gmac** [comment]
For posterity, this PR now has a shoutout in the GraphQL Tools documentation here: https://www.graphql-tools.com/docs/batch-execution#batch-execution. The GraphQL Tools implementation is more generic, but is unflappably based on this excellent PR.
